server/src/main/java/org/elasticsearch/common/LocalTimeOffset.java

```diff
@@ -0,0 +1,679 @@
+        protected final long minUtcMillis;
+        protected final long maxUtcMillis;
+
+        AbstractManyTransitionsLookup(ZoneId zone, long minUtcMillis, long maxUtcMillis) {
+            this.zone = zone;
+            this.minUtcMillis = minUtcMillis;
+            this.maxUtcMillis = maxUtcMillis;
+        }
+
+        @Override
+        public final LocalTimeOffset lookup(long utcMillis) {
+            assert utcMillis >= minUtcMillis;
+            assert utcMillis <= maxUtcMillis;
+            return innerLookup(utcMillis);
+        }
+
+        protected abstract LocalTimeOffset innerLookup(long utcMillis);
+
+        @Override
+        public final LocalTimeOffset fixedInRange(long minUtcMillis, long maxUtcMillis) {
+            LocalTimeOffset offset = lookup(maxUtcMillis);
+            return offset.containsUtcMillis(minUtcMillis) ? offset : null;
+        }
+
+        protected static NoPrevious buildNoPrevious(ZoneOffsetTransition transition) {
+            return new NoPrevious(transition.getOffsetBefore().getTotalSeconds() * 1000);
+        }
+
+        protected static Transition buildTransition(ZoneOffsetTransition transition, LocalTimeOffset previous) {
+            long utcStart = transition.toEpochSecond() * 1000;
+            long offsetBeforeMillis = transition.getOffsetBefore().getTotalSeconds() * 1000;
+            long offsetAfterMillis = transition.getOffsetAfter().getTotalSeconds() * 1000;
+            if (transition.isGap()) {
+                long firstMissingLocalTime = utcStart + offsetBeforeMillis;
+                long firstLocalTimeAfterGap = utcStart + offsetAfterMillis;
+                return new Gap(offsetAfterMillis, previous, utcStart, firstMissingLocalTime, firstLocalTimeAfterGap);
+            }
+            long firstOverlappingLocalTime = utcStart + offsetAfterMillis;
+            long firstNonOverlappingLocalTime = utcStart + offsetBeforeMillis;
+            return new Overlap(offsetAfterMillis, previous, utcStart, firstOverlappingLocalTime, firstNonOverlappingLocalTime);
+        }
+    }
+
+    private static LocalTimeOffset checkForFixedZone(ZoneId zone, ZoneRules rules) {
+        if (false == rules.isFixedOffset()) {
+            return null;
+        }
+        LocalTimeOffset fixedTransition = new NoPrevious(rules.getOffset(Instant.EPOCH).getTotalSeconds() * 1000);
+        return fixedTransition;
+    }
+
+    /**
+     * The maximum number of {@link ZoneOffsetTransition} to collect before
+     * giving up because the date range will be "too big". I picked this number
+     * fairly arbitrarily with the following goals:
+     * <ol>
+     * <li>Don't let {@code lookup(Long.MIN_VALUE, Long.MAX_VALUE)} consume all
+     *     the memory in the JVM.
+     * <li>It should be much larger than the number of offsets I'm bound to
+     *     collect.
+     * </ol>
+     * {@code 5_000} collects about 2_500 years worth offsets which feels like
+     * quite a few!
```

Comment:
It's also likely completely invalid; surely we'll have abolished DST at some point in the next 2500 years? :grin:

How well are we testing the case where we hit this limit?

Reply:
> It's also likely completely invalid; surely we'll have abolished DST at some point in the next 2500 years? grin

Gosh I hope so!



> How well are we testing the case where we hit this limit?

We test the simple case of not having a lookup and it is simple. We test the more complex side - what we do when we don't have a lookup - in a way that isn't super satisfying now that I think about it. We build something with a limit a useful lookup and then build a second version forcing the `java.time` implementation and "duel" them. I do think it'd be worth preparing a couple of roundings with stupidly large ranges and checking a few things about them. At this point I'd be ok with spot checking a few roundings and smashing a bunch of random dates at it to make sure it doesn't crash.

Reply:
I pushed some test cases with explicit far future dates.